src/org/rascalmpl/repl/TerminalProgressBarMonitor.java

```diff
@@ -117,87 +118,80 @@ public void write(String str, int off, int len) throws IOException {
         }
 
         public void write(byte[] n, OutputStream out) throws IOException {
-            add(n, 0, n.length);
-            flushToLastLine(out);
+            write(n, 0, n.length, out);
         }
 
+        /**
+         * Main workhorse looks for newline characters in the new input.
+         *  - if there are newlines, than whatever is in the buffer can be flushed.
+         *  - all the characters up to the last new new line are flushed immediately.
+         *  - all the new characters after the last newline are buffered.
+         */
         public void write(byte[] n, int offset, int len, OutputStream out) throws IOException {
-            add(n, offset, len);
-            flushToLastLine(out);
-        }
-
-        private void flushToLastLine(OutputStream out) throws IOException {
-            if (lastNewLine != -1) {
-                // write everything out (except the last unfinished line), but including the last newline
-                out.write(buffer, 0, lastNewLine + 1);
-                
-                if (lastNewLine + 1 == curEnd) {
-                    // nothing left
-                    curEnd = 0;
-                    lastNewLine = -1;
-                }
-                else {
-                    // copy the last line that does not end with a newline
-                    curEnd -= lastNewLine;
-                    System.arraycopy(buffer, lastNewLine + 1, buffer, 0, curEnd);
-                    lastNewLine = -1; // no newline anymore
-                }
-            }
+            int lastNL = startOfLastLine(n, offset, len);
 
-            // otherwise we wait until the next input comes to be able to complete a line.
+            if (lastNL == -1) {
+                store(n, offset, len);
+            }
+            else {
+                flush(out);
+                out.write(n, offset, lastNL + 1);
+                store(n, offset, lastNL - offset);
+            }
         }
 
-        public void writeLeftOvers(OutputStream out) throws IOException {
+        /**
+         * This empties the current buffer onto the stream,
+         * and resets the cursor.
+         */
+        private void flush(OutputStream out) throws IOException {
             if (curEnd != 0) {
                 out.write(buffer, 0, curEnd);
-                out.write('\n');
                 curEnd = 0;
-                lastNewLine = -1;
+            }
+        }
+
+        /**
+         * Prints whatever is the last line in the buffer,
+         * and adds a newline.
+         */
+        public void flushLastLine(OutputStream out) throws IOException {
+            if (curEnd != 0) {
+                flush(out);
+                out.write('\n');
             }
         }
     
-        private int startOfLastLine() {
-            for (int i = curEnd - 1; i >= 0; i--) {
+        private int startOfLastLine(byte[] buffer, int offset, int len) {
+            for (int i = offset + len - 1; i >= 0; i--) {
```

Comment:
This should stop at offset, not at 0.

Reply:
Nice catch!

Reply:
fix with `--author davy` . I think there are some more issues though.. I'll fix them.